<a href="https://colab.research.google.com/github/PosgradoMNA/ActividadesAprendizaje-A01794419-IgnacioQuintero/blob/main/Module2_Data_Analysis_with_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Analysis with Python
## Module 2 - Data Wrangling

* Alumno: Ignacio Antonio Quintero Chávez.
* Matrícula: A01794419.
* Materia: Ciencia y analítica de datos.
* Profesor: Jobish Vallikavungal Devassia.
* Fecha: 18 Octubre 2022.

**Data Pre-processing (Data Cleaning or Data Wrangling)**

The process of converting or mapping data from the initial "raw" form into another format, in order to prepare the data for further analysis.

### Import data

In [1]:
# Importing the libraries
import pandas as pd
import matplotlib.pylab as plt

In [2]:
# Importing the dataset
dataset = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"

In [3]:
# Creating a python list headers containing name of headers
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [5]:
dataframe = pd.read_csv(dataset, names = headers)

In [6]:
# Display the first five rows of the dataframe
dataframe.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

Steps for working with missing data:

* Identify missing data.
* Deal with missing data.
* Correct data format.



### Identify and handle missing values

**Convert "?" to NaN**

Using the function 

.replace(A, B, inplace = True)

In [8]:
import numpy as np

# Replace "?" to NaN
dataframe.replace("?", np.nan, inplace = True)
dataframe.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

All the values '?' were replaced with NaN (Not a Number).

**Evaluating for missing data**

The missing values are converted by default. There are two methods to detect missing data:

* isnull()
* notnull()

In [9]:
missing_data = dataframe.isnull()
missing_data.head()

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

**True** means the value is a missing value while **False** means the value is not a missing value.

**Count missing values in each column**

Using a for loop in Python allow us to figure out the number of missing values in each column. **True** represents a missing value and **False** means the value is present in the dataset.

The method value_counts() counts the number of True values.

In [10]:
for column in missing_data.columns.values.tolist():
  print(column)
  print(missing_data[column].value_counts())
  print("")

symboling
False    205
Name: symboling, dtype: int64

normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64

make
False    205
Name: make, dtype: int64

fuel-type
False    205
Name: fuel-type, dtype: int64

aspiration
False    205
Name: aspiration, dtype: int64

num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64

body-style
False    205
Name: body-style, dtype: int64

drive-wheels
False    205
Name: drive-wheels, dtype: int64

engine-location
False    205
Name: engine-location, dtype: int64

wheel-base
False    205
Name: wheel-base, dtype: int64

length
False    205
Name: length, dtype: int64

width
False    205
Name: width, dtype: int64

height
False    205
Name: height, dtype: int64

curb-weight
False    205
Name: curb-weight, dtype: int64

engine-type
False    205
Name: engine-type, dtype: int64

num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64

engine-size
False    205
Name: engine-size, dtype: int64

fuel-system
Fa

Based on the summary above, each column has 205 rows of data and seven of the columns containing missing data:


* normalized-losses: 41 missing data.
* num-of-doors: 2 missing data.
* bore: 4 missing data.
* stroke: 4 missing data.
* horsepower: 2 missing data.
* peak-rpm: 2 missing data.
* price: 4 missing data.


**Deal with missing data**

1. Drop data.

* Drop the whole row.
* Drop the whole column.

2. Replace data.
* Replace it by mean.
* Replace it by frequency.
* Replace it based on other functions.

**Calculate the mean value for the "normalized-losses" column.**

In [11]:
avg_norm_loss = dataframe["normalized-losses"].astype("float").mean(axis=0)
print("The average of normalized-losses is:", avg_norm_loss)

The average of normalized-losses is: 122.0


**Replace "NaN" with mean value in "normalized-losses" column.**

In [12]:
dataframe["normalized-losses"].replace(np.nan, avg_norm_loss, inplace = True)

**Calculate the mean value for the "bore" column.**

In [13]:
avg_bore = dataframe["bore"].astype("float").mean(axis=0)
print("The average of bore is:", avg_bore)

The average of bore is: 3.3297512437810943


**Replace "NaN" with the mean value in the "bore" column.**

In [14]:
dataframe["bore"].replace(np.nan, avg_bore, inplace = True)

**Calculate the mean value for stroke column.**

In [15]:
avg_stroke = dataframe["stroke"].astype("float").mean(axis = 0)
print("The average of stroke is:", avg_stroke)

# replace NaN by mean value in "stroke" column
dataframe["stroke"].replace(np.nan, avg_stroke, inplace = True)

The average of stroke is: 3.255422885572139


**Calculate the mean value for the horsepower column.**

In [16]:
avg_horsepower = dataframe['horsepower'].astype('float').mean(axis=0)
print("The average horsepower is:", avg_horsepower)

The average horsepower is: 104.25615763546799


**Replace NaN with the mean value in the horsepower column.**

In [17]:
dataframe['horsepower'].replace(np.nan, avg_horsepower, inplace=True)

**Calculate the mean value for peak-rpm column.**

In [18]:
avg_peakrpm=dataframe['peak-rpm'].astype('float').mean(axis=0)
print("The average peak rpm is:", avg_peakrpm)

The average peak rpm is: 5125.369458128079


**Replace NaN with the mean value in the peak-rpm column.**

In [19]:
dataframe['peak-rpm'].replace(np.nan, avg_peakrpm, inplace=True)

In [20]:
# Show which values are present in a particular column
dataframe['num-of-doors'].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

In [21]:
# Calculating the most common type automatically
dataframe['num-of-doors'].value_counts().idxmax()

'four'

In [22]:
# Drop all the rows that don't have price data

# simply drop whole row with NaN in "price" column
dataframe.dropna(subset=["price"], axis=0, inplace=True)

# reset index, because we droped two rows
dataframe.reset_index(drop=True, inplace=True)

In [24]:
dataframe.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]